src/spi.rs

```diff
@@ -779,6 +779,65 @@ macro_rules! hal {
                     Spi { spi, pins }
                 }
 
+                /// Enable transmit register empty interrupt
+                pub fn enable_txe_interrupt(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() | (1 << 7))});
+                }
+
+                /// Disable transmit register empty interrupt
+                pub fn disable_txe_interrupt(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() & !(1 << 7))});
+                }
+
+                /// Enable receive register not empty interrupt
+                pub fn enable_rxne_interrupt(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() | (1 << 6))});
+                }
+
+                /// Disable receive register not empty interrupt
+                pub fn disable_rxne_interrupt(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() & !(1 << 6))});
+                }
+
+                /// Enable error interrupt
+                pub fn enable_error_interrupt(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() | (1 << 5))});
+                }
+
+                /// Disable error interrupt
+                pub fn disable_error_interrupt(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() & !(1 << 5))});
+                }
+
+                /// Convenience function to enable all interrupts
+                pub fn enable_interrupts(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() | (7 << 5))});
+                }
+
+                /// Convenience function to disable all interrupts
+                pub fn disable_interrupts(&mut self) {
+                    self.spi.cr2.modify(|r, w| unsafe {w.bits(r.bits() & !(7 << 5))});
+                }
+
+                /// Enable the peripheral, i.e. set the SPE bit in the SPI control register 1
+                pub fn enable(&mut self) {
+                    self.spi.cr1.modify(|r, w| unsafe {w.bits(r.bits() | (1 << 6))});
+                }
+
+                /// Halt the SPI according to the procedure described in chapter
+                /// 28.3.8 of the STM32F4xx reference manual
+                // TODO: this only applies to master bi-directional transfer mode
+                pub fn wait_until_idle(&mut self) {
+                    while self.spi.sr.read().rxne().bit_is_clear() {}
```

Comment:
The more traditional way here would be to only have one loop, one read and three checks in one go breaking the loop when all conditions are fulfilled.

Reply:
The STM32f4 Reference Manual implies an order to the checks:
![manual](https://user-images.githubusercontent.com/9446963/54075115-99d8d800-429b-11e9-8c8c-0cd37d97db5e.png)

I also think another function that allows skipping the `rxne` check might be handy. I see an issue with applications that need to read the received data before they know if the transfer is complete. With the API as it is in the PR, this would go into an infinite loop:

```rust
let data = spi.read();    // clears rxne bit
if transfer_done(data) {
    spi.wait_until_idle(); // rxne cleared at this point -> infinite loop
    spi.disable();
}
```
What do you think?

Reply:
It is weird that there's a fixed order with separate readings here especially since those events could trigger again in the meantime. I didn't quite get the point of `wait_for_idle()`: Ideally we don't have to use it at all unless we want to disable the peripheral, but in that case it really should be done automatically to prevent usage errors.

Reply:
Ok, I agree. How about I remove `enable()`, `disable()`, and `wait_until_idle()` from this PR alltogether and we see if we need it in the future?

Reply:
Works for me.